R/two.step.hybrid.R

```diff
@@ -1,170 +1,397 @@
+  metadata <- read.table(metadata, sep = ",", header = TRUE)
+  filenames_batchwise <- bind_batch_label_column(filenames, metadata)
+  batches_idx <- unique(metadata$batch)
+  batchwise <- new("list")
+  message("* processing ", length(batches_idx), " batches separately")
+
+  for (batch.i in 1:length(batches_idx)) {
+    files_batch <- dplyr::filter(filenames_batchwise, batch == batch.i)$filename
+    message("* processing ", length(files_batch), " samples from batch ", batch.i)
+
+    features <- semi.sup(
+      files = files_batch,
+      folder = work_dir,
+      n.nodes = cluster,
+      known.table = known.table,
+      sd.cut = sd.cut,
+      sigma.ratio.lim = sigma.ratio.lim,
+      component.eliminate = component.eliminate,
+      moment.power = moment.power,
+      min.pres = min.pres,
+      min.run = min.run,
+      min.exp = ceiling(min.within.batch.prop.detect * length(files_batch)),
+      mz.tol = mz.tol,
+      baseline.correct.noise.percentile = baseline.correct.noise.percentile,
+      baseline.correct = baseline.correct,
+      align.mz.tol = align.mz.tol,
+      align.chr.tol = align.chr.tol,
+      max.align.mz.diff = max.align.mz.diff,
+      recover.mz.range = recover.mz.range,
+      recover.chr.range = recover.chr.range,
+      use.observed.range = use.observed.range,
+      shape.model = shape.model,
+      new.feature.min.count = new.feature.min.count,
+      recover.min.count = recover.min.count
+    )
+
+    features$final.ftrs <- features$final.ftrs[order(features$final.ftrs[, 1], features$final.ftrs[, 2]), ]
+    features$final.times <- features$final.times[order(features$final.times[, 1], features$final.times[, 2]), ]
+
+    batchwise[[batch.i]] <- features
+  }
+
+  batchwise <- batchwise_wrapper(batchwise, batches_idx)
+
+  step_one_features <- list()
+  for (batch_id in batches_idx) {
+    step_one_features[[batch_id]] <- compute_intensity_medians(
+      batchwise[[batch_id]]$recovered_feature_sample_table
+    ) %>%
+      dplyr::select(-c("mz_min", "mz_max"))
+  }
+
+  cl <- makeCluster(cluster)
+  registerDoParallel(cl)
+  clusterEvalQ(cl, library(recetox.aplcms))
+
+  message("* aligning time")
+  corrected <- adjust.time(step_one_features,
+    mz.tol = batch.align.mz.tol,
+    chr.tol = batch.align.chr.tol,
+    find.tol.max.d = 10 * mz.tol,
+    max.align.mz.diff = max.align.mz.diff,
+    rt_colname = "rt"
+  )
+
+  message("* aligning features")
+  aligned <- align_features(
+    sample_names = paste0("batch_", batches_idx),
+    features = corrected,
+    min.exp = ceiling(min.batch.prop * length(batches_idx)),
+    mz.tol = batch.align.mz.tol,
+    chr.tol = batch.align.chr.tol,
+    find.tol.max.d = 10 * mz.tol,
+    max.align.mz.diff = max.align.mz.diff,
+    rt_colname = "rt"
+  )
+
+  aligned <- as_wide_aligned_table(aligned)
+
+  stopCluster(cl)
+
+  message("* recovering features across batches")
+
+  cl <- makeCluster(cluster)
+  registerDoParallel(cl)
```

Comment:
Cluster is stopped and then created again. Is there any reason for that? Also, are we sure the cluster is actually being used here?

Reply:
Yeah, seems like the cluster is just never used - I think it just can be deleted.

Reply:
It is used in `adjust.time`, which uses `%dopar%`. I think you're right though, in that there's no need for creating and closing the cluster multiple times. Fixed that in 527f01c.